### PR TITLE
ConcurrentModificationException bug fixed

### PR DESCRIPTION
**Pull request by avpotapov00**
Due to incorrect LoopDetector work, we ran into an infinite spin-cycle on trace collection. The test failed with timeout but in test threads, we continued to add elements to the trace list. That caused ConcurrentModificationException when we tried to traverse it to create this trace representation.

### CODE REVIEW

File: src/jvm/test/resources/expected_logs/obstruction_freedom_violation_with_no_detected_cycle.txt [hunk 1]
```diff
@@ -127,6 +126,5 @@ Detailed trace:
 |     incrementAndGet(): 101 at ObstructionFreedomActiveLockRepresentationTest.incrementManyTimes(ObstructionFreedomRepresentationTest.kt:44) |
 |   incrementManyTimes() at ObstructionFreedomActiveLockRepresentationTest.operation(ObstructionFreedomRepresentationTest.kt:37)              |
 |     get(): 101 at ObstructionFreedomActiveLockRepresentationTest.incrementManyTimes(ObstructionFreedomRepresentationTest.kt:41)             |
-|     incrementAndGet(): 102 at ObstructionFreedomActiveLockRepresentationTest.incrementManyTimes(ObstructionFreedomRepresentationTest.kt:44) |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/obstruction_freedom_violation_with_no_detected_cycle.txt:130]: This event should be in the trace, shouldn't it?

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    @Operation
+    fun enqueue(@Param(name = "element") element: Int) = queue.enqueue(element)
+
+    @Operation
+    fun dequeue() = queue.dequeue() != null
+
+    @Test
+    fun modelCheckingTest() =
+        ModelCheckingOptions()
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt:40]: Can you simplify the test configuration?

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            .checkObstructionFreedom(false)
+            .minimizeFailedScenario(false)
+            .sequentialSpecification(BoundedIntQueueSequential::class.java)
+            .checkImpl(this::class.java)
+            .checkLincheckOutput("bounded_queue_incorrect_results.txt")
+
+}
+
+@Suppress("unused")
+class BoundedIntQueueSequential {
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt:56]: I don't think you need it

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import org.jetbrains.kotlinx.lincheck.paramgen.IntGen
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingOptions
+import org.jetbrains.kotlinx.lincheck_test.util.checkLincheckOutput
+import org.junit.Test
+
+/**
+ * This test is created to verify this [bug](https://github.com/JetBrains/lincheck/issues/209) is resolved.
+ * The problem was that we ran in infinite spin-lock in trace collection mode due to incorrect Loop detector work.
+ */
+@Param(name = "element", gen = IntGen::class, conf = "0:3")
```
Comment (ndkoval) [inline src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt:28]: You don't need it

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 5]
```diff
@@ -276,25 +276,69 @@ abstract class ManagedStrategy(
         if (!isTestThread(iThread)) return // can switch only test threads
         if (inIgnoredSection(iThread)) return // cannot suspend in ignored sections
         check(iThread == currentThread)
-        var isLoop = false
-        if (loopDetector.visitCodeLocation(iThread, codeLocation)) {
+
+        if (loopDetector.replayModeEnabled) {
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:280]: Please clarify why the logic should be different

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 6]
```diff
@@ -276,25 +276,69 @@ abstract class ManagedStrategy(
...
+            } else {
+                switchCurrentThread(iThread, SwitchReason.STRATEGY_SWITCH)
+            }
+        } else {
+            loopDetector.onNextExecutionPoint(codeLocation)
+        }
+    }
+
+    /**
+     * When replaying executions it's important to repeat exactly the same executions and switches,
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:316]: This comment has some grammar issues, please use Grammarly to fix them

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt (hunk 2 above)
Comment (avpotapov00) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt:40]: I removed the setting and replaced it with a custom scenario configuration. I think it's a good approach to use in the test exactly the same scenario as caused the issue.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt (hunk 3 above)
Comment (ndkoval) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt:70]: Why do you need it?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 7]
```diff
@@ -276,25 +276,73 @@ abstract class ManagedStrategy(
...
+             for the next 200-12 operations. This leads to the results of another execution, compared to the
+             original failure results.
+             To avoid this bug when we're replaying some executions, we have to follow only loopDetector history during
+             the last execution. In the considered example, we will retain that we will switch soon after
+             the spin cycle in thread 1, so no bug will appear.
+             */
+            newSwitchPointInReplayMode(iThread, codeLocation, tracePoint)
+        } else {
+            /*
+            In the regular mode we just have listen both interleaving and loopDetector to determine should we
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:298]: > we just have listen both interleaving and loopDetector

please clarify

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 8]
```diff
@@ -895,15 +955,29 @@ abstract class ManagedStrategy(
         }
 
         private fun onNextThreadSwitchPoint(nextThread: Int) {
+            /*
+                When we're back to some thread, newSwitchPoint won't be called before the fist
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:959]: before the first EVENT?

File: src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt [hunk 9]
```diff
@@ -895,15 +955,29 @@ abstract class ManagedStrategy(
...
+                in current thread part as it was called before switch.
```
Comment (ndkoval) [inline src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ManagedStrategy.kt:960]: Please re-read your changes and comments slowly, so you can polish them. Then, the PR should be ready to go.

**Review by ndkoval — CHANGES_REQUESTED** [review]

File: src/jvm/test/resources/expected_logs/bounded_queue_incorrect_results.txt [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+|                                                                                             |     concurrentEnqueues.READ: 1 at QueueAdaptor.dequeue(BoundedQueueTest.kt:105) |            |
+|                                                                                             |     dequeue(): null at QueueAdaptor.dequeue(BoundedQueueTest.kt:99)             |            |
+|                                                                                             |     concurrentEnqueues.READ: 1 at QueueAdaptor.dequeue(BoundedQueueTest.kt:105) |            |
+|                                                                                             |     dequeue(): null at QueueAdaptor.dequeue(BoundedQueueTest.kt:99)             |            |
+|                                                                                             |     concurrentEnqueues.READ: 1 at QueueAdaptor.dequeue(BoundedQueueTest.kt:105) |            |
+|                                                                                             |     dequeue(): null at QueueAdaptor.dequeue(BoundedQueueTest.kt:99)             |            |
```
Comment (ndkoval) [inline src/jvm/test/resources/expected_logs/bounded_queue_incorrect_results.txt:117]: You should've detected spin-lock here.

Comment (avpotapov00) [reply src/jvm/test/resources/expected_logs/bounded_queue_incorrect_results.txt:117]: The problem was that previously I didn't decrease `Interleaving.executionPosition`, so after a cycle was detected for the first time we got a lot of new potential switch positions with a big execution number. 
Now `Interleaving` is rolled back every time we find spin-lock so potential switch points stay consistent.

File: src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt (hunk 3 above)
Comment (avpotapov00) [reply src/jvm/test/org/jetbrains/kotlinx/lincheck_test/representation/BoundedQueueTest.kt]: Finally removed, somehow didn't notice it twice.

**Review by ndkoval — APPROVED** [review]

**Pull request merged**